[WIP] Consolidated 4.10 RAN docs (#43890)
openshift/openshift-docs: modules/cnf-topology-aware-lifecycle-operator-operator-update.adoc

====

.. Verify that the subscription policy is `NonCompliant` at this point by running the following command:
+
[source,terminal]
----
$ oc get policy common-subscriptions-policy -n <policy_namespace>
----
+
.Example output
+
[source,terminal]
----
NAME                        REMEDIATION ACTION   COMPLIANCE STATE     AGE
common-subscriptions-policy   inform               NonCompliant       27d
----

.. Apply the `cgu-operator-upgrade.yaml` to the Hub cluster by running the following command:
+
[source,terminal]
----
$ oc apply -f cgu-operator-upgrade.yml
----

.. Monitor the process and wait for the pre-caching to complete. Check the status of pre-caching by running the following command on the spoke cluster:
+
[source,terminal]
----
$ oc get jobs,pods -n openshift-talo-pre-cache
----

.. Check if the pre-caching is completed before starting the update by running the following command:
+
[source,terminal]
----
$ oc get cgu cgu-upgrade-complete -ojson | jq '.status.conditions'
----
+
.Example output
+

> [source,json]

Reviewer: output is like this:

```
[
  {
    "lastTransitionTime": "2022-04-08T14:14:40Z",
    "message": "The ClusterGroupUpgrade CR has all clusters already compliant with the specified managed policies",
    "reason": "UpgradeCompleted",
    "status": "False",
    "type": "Ready"
  }
]
```

kind: PolicyGenTemplate
metadata:
  name: "du-upgrade"
  namespace: "ztp-group-du-sno"
spec:
  bindingRules:
    group-du-sno: ""
  mcp: "master"
  remediationAction: inform
  sourceFiles:
    - fileName: DefaultCatsrc.yaml
      remediationAction: inform
      policyName: "operator-catsrc-policy"
      metadata:
        name: redhat-operators
      spec:
        displayName: Red Hat Operators Catalog
        image: registry.example.com:5000/olm/redhat-operators:v4.10 <1>
        updateStrategy: <2>
          registryPoll:
            interval: 10m
----
<1> The index image URL contains the desired Operator images. If the index images are always pushed to the same image name and tag, this change is not needed.
<2> Set how frequently Operator Lifecycle Manager (OLM) pulls the index image for new Operator versions with the `registrypoll.interval` field. This change is not needed if a new index image tag is always pushed for y-stream and z-stream Operator updates.

.. This update generates one policy, `du-upgrade-operator-catsrc-policy`, to update the `redhat-operators` catalog source with the new index images that contain the desired Operators images.
+

> [NOTE]

Author: @yliu127 @Missxiaoguo Could you please check if the rephrasing in this note correct? It was a bit too conceptual in a procedural flow, and the peer reviewer asked me to be more action focused.
@bobfuru Please re-review :)

Original if needed:
> If the desired Operators are not from one catalog source, all catalog sources must be properly updated with the new index image or registry poll interval. To use the image pre-caching for Operators later on, the updates for the different catalog sources must be applied in different policies. In addition, the separated subscription policies are required for the desired Operators from different catalog sources.

====
When you want to update Operators from different catalog sources, you must do the following:
* Update all catalog sources with the new index image or registry poll interval.
* To use the image pre-caching for Operators later on, apply the updates for the different catalog sources in different policies.

> * Prepare separate subscription policies for the desired Operators from different catalog sources.

Reviewer 2: ```suggestion
If you will use the image pre-caching for Operators and there are Operators from a different catalog source other than **redhat-operators**,  you must perform the following tasks:
* Prepare a separate catalog source policy with the new index image or registry poll interval update for the different catalog source
* Prepare a separate subscription policy for the desired Operators that are from the different catalog source
```

  sourceFiles:
    - fileName: DefaultCatsrc.yaml
      remediationAction: inform
      policyName: "operator-catsrc-policy"
      metadata:
        name: redhat-operators
      spec:
        displayName: Red Hat Operators Catalog
        image: registry.example.com:5000/olm/redhat-operators:v4.10 <1>
        updateStrategy: <2>
          registryPoll:
            interval: 10m
----
<1> The index image URL contains the desired Operator images. If the index images are always pushed to the same image name and tag, this change is not needed.
<2> Set how frequently Operator Lifecycle Manager (OLM) pulls the index image for new Operator versions with the `registrypoll.interval` field. This change is not needed if a new index image tag is always pushed for y-stream and z-stream Operator updates.

.. This update generates one policy, `du-upgrade-operator-catsrc-policy`, to update the `redhat-operators` catalog source with the new index images that contain the desired Operators images.
+
[NOTE]
====
If you want to use the image pre-caching for Operators and there are Operators from a different catalog source other than `redhat-operators`,  you must perform the following tasks:

> * Update all catalog sources with the new index image or registry poll interval.
> * To use the image pre-caching for Operators later on, apply the updates for the different catalog sources in different policies.
> * Prepare separate subscription policies for the desired Operators from different catalog sources.

Reviewer 2: Just a reminder. This was not updated correctly

```suggestion
* Prepare a separate catalog source policy with the new index image or registry poll interval update for the different catalog source
* Prepare a separate subscription policy for the desired Operators that are from the different catalog source
```

+
ArgoCD pulls the changes from the Git repository and generates the policies on the hub cluster.

.. Check the created policies by running the following command:
+
[source,terminal]
----
$ oc get policies -A | grep -E "catsrc-policy|subscription"
----

. Apply the required catalog source updates before starting the Operator update.

.. Save the content of the `ClusterGroupUpgrade` CR named `operator-upgrade-prep` with the catalog source policies and the target spoke clusters to the `cgu-operator-upgrade-prep.yml` file:
+
[source,yaml]
----
apiVersion: ran.openshift.io/v1alpha1
kind: ClusterGroupUpgrade
metadata:
  name: cgu-operator-upgrade-prep
  namespace: default
spec:
  clusters:
  - spoke1
  enable: true
  managedPolicies:
  - du-upgrade-operator-catsrc-policy
  remediationStrategy:
    maxConcurrency: 1
----

.. Apply the policy to the hub cluster by running the following command:
+
[source,terminal]
----
$ oc apply -f cgu-operator-upgrade-prep.yml
----

.. Monitor the update process. Upon completion, ensure that the policy is compliant by running the following command:
+
[source,terminal]
----
$ oc get policies -A | grep -E "catsrc-policy"
----

> . Create the `ClusterGroupUpdate` CR for the Operator update with the `spec.enable` field set to `false`.
> .. Save the content of the Operator update `ClusterGroupUpdate` CR with the `du-upgrade-operator-catsrc-policy` policy and the subscription policies created from the common `PolicyGenTemplate` and the target clusters to the `cgu-operator-upgrade.yml` file, as shown in the following example:

Reviewer 2: ```suggestion
. Create the `ClusterGroupUpgrade` CR for the Operator update with the `spec.enable` field set to `false`.
.. Save the content of the Operator update `ClusterGroupUpgrade` CR with the `du-upgrade-operator-catsrc-policy` policy and the subscription policies created from the common `PolicyGenTemplate` and the target clusters to the `cgu-operator-upgrade.yml` file, as shown in the following example:
```

  managedPolicies:
  - du-upgrade-operator-catsrc-policy <1>
  - common-subscriptions-policy <2>
  preCaching: false
  clusters:
  - spoke1
  remediationStrategy:
    maxConcurrency: 1
  enable: false
----
<1> The policy updates the catalog source.
<2> The policy contains operator subscriptions. If you have upgraded ZTP from 4.9 to 4.10 by following "Upgrade ZTP from 4.9 to 4.10", all Operator subscriptions are grouped into the `common-subscriptions-policy policy`.
+
[NOTE]
====
One `ClusterGroupUpgrade` CR can only pre-cache the images of the desired Operators defined in the subscription policy from one catalog source included in the `ClusterGroupUpgrade` CR. If the desired Operators are from different catalog sources, such as in the example of the SRIOV-FEC Operator, another `ClusterGroupUpgrade` CR must be created with `du-upgrade-fec-catsrc-policy` and `du-upgrade-subscriptions-fec-policy` policies for the SRIOV-FEC Operator images pre-caching and update.
====

.. Apply the `ClusterGroupUpgrade` CR to the hub cluster by running the following commands:
+
[source,terminal]
----

> $ oc apply -f cgu-platform-upgrade.yml

Reviewer: should be cgu-operator-upgrade.yml

+
[source,terminal]
----
$ oc get cgu -n default cgu-upgrade-complete -ojsonpath='{.status.conditions}'
----
+

> .Example output

Reviewer: The example output does not match the context. This is an out for upgrade completion, while we are expecting an output that indicates precaching completed (but upgrade has not started).

Reviewer: This looks good to me.

Author: Updated
